### ISSUE TITLE
REST API Pagination Does not work as expected

### ISSUE THREAD

**Issue by lbajsarowicz** [magento/magento2#26986]
### Preconditions (*)
1. Magento 2.3.4
2. REST API Call

### Steps to reproduce (*)
1. Fixture with 3 orders
2. Request with `page_size` = 1 and `current_page` = 99

### Expected result (*)
1. `total_count` returns 3
2. `items` returns empty array **OR** OutOfRangeException

### Actual result (*)
1. `total_count` is 3
2. `items` returns repeated last page of results

### Kudos for
@ihor-sviziev , @Vinai , @nuzil for discussion at `#appdesign` channel about this issue.
![image](https://user-images.githubusercontent.com/1639941/75116803-46a69380-566c-11ea-8ed7-4dfc33b6acdf.png)
![image](https://user-images.githubusercontent.com/1639941/75116807-502ffb80-566c-11ea-9cbf-c08a9321fa75.png)
![image](https://user-images.githubusercontent.com/1639941/75116809-545c1900-566c-11ea-9b4d-6643f0ae1857.png)

**Comment by lbajsarowicz** [comment]
I'm going to address the issue with data described with `dataProvider`:
```php

    /**
     * Keep in mind: Fixture contains 5 products
     *
     * @return array
     */
    public function productPaginationDataProvider()
    {
        return [
            'expect-all-items' => [
                'pageSize' => 10,
                'currentPage' => 1,
                'expectedCount' => 5
            ],
            'expect-page=size-items' => [
                'pageSize' => 2,
                'currentPage' => 1,
                'expectedCount' => 2
            ],
            'expect-less-than-pagesize-elements' => [
                'pageSize' => 3,
                'currentPage' => 2,
                'expectedCount' => 2
            ],
            'expect-no-items' => [
                'pageSize' => 100,
                'currentPage' => 99,
                'expectedCount' => 0
            ]
        ];
    }
```

**Comment by lbajsarowicz** [comment]
The source of the issue is located in:

https://github.com/magento/magento2/blob/09de30c806de5f69c081991f8afabcdfc3b19926/lib/internal/Magento/Framework/Data/Collection.php#L240-L249

Although `$this->_curPage` is overwritten by the 

```php
    if ($this->_curPage + $displacement > $this->getLastPageNumber()) {
        return $this->getLastPageNumber();
    }
```

That is not passed back to all APIs. There are two options, actually:
1. Remove above-mentioned mechanism to return no values when we reached the last page.
2. Pass the information about overwritten `currentPage` to API

I choose (1) as for me it's more obvious.

**Comment by magento-engcom-team** [comment]
Hi @lbajsarowicz. Thank you for your report. 
The issue has been fixed in magento/magento2#26988 by @lbajsarowicz in 2.4-develop branch
Related commit(s):
 - [a6f76f346596a80243596079f90fe0b78f7dc744](https://github.com/magento/magento2/commit/a6f76f346596a80243596079f90fe0b78f7dc744)
 - [b254764a3c654b8ed48c36086864478bfb6c953e](https://github.com/magento/magento2/commit/b254764a3c654b8ed48c36086864478bfb6c953e)
 - [11567f6004dbc6819c19bd42e95915dfbf3a046f](https://github.com/magento/magento2/commit/11567f6004dbc6819c19bd42e95915dfbf3a046f)
 - [6771122e05e1c49fc4f83409bd6fd18ffaaf0951](https://github.com/magento/magento2/commit/6771122e05e1c49fc4f83409bd6fd18ffaaf0951)
 - [e001493a7877e5f00a53ae1d34a98039b8e0313e](https://github.com/magento/magento2/commit/e001493a7877e5f00a53ae1d34a98039b8e0313e)
 - [ecb3a8d393854ab1ab0543f541fd64d5c1941015](https://github.com/magento/magento2/commit/ecb3a8d393854ab1ab0543f541fd64d5c1941015)
 - [015c5fcc53b96efa460c05ec4047006fbf68e753](https://github.com/magento/magento2/commit/015c5fcc53b96efa460c05ec4047006fbf68e753)
 - [18e43a8df8581586b7ba3110e3feebcd6e1910d7](https://github.com/magento/magento2/commit/18e43a8df8581586b7ba3110e3feebcd6e1910d7)
 - [9d0ae13e97e6403766fe7eba7cde9946095866e5](https://github.com/magento/magento2/commit/9d0ae13e97e6403766fe7eba7cde9946095866e5)
 - [2439e17df3ec42440f25efad2e2b2a7d3be0de5e](https://github.com/magento/magento2/commit/2439e17df3ec42440f25efad2e2b2a7d3be0de5e)
 - [14ea3358e08c26403d1a322f270daa55a4d2e033](https://github.com/magento/magento2/commit/14ea3358e08c26403d1a322f270daa55a4d2e033)
 - [0fe9cf207bc902350d5ebfc4b5a665679db6edc0](https://github.com/magento/magento2/commit/0fe9cf207bc902350d5ebfc4b5a665679db6edc0)
 - [dab0cda47e62d8b519fcac798f9221066615d7ec](https://github.com/magento/magento2/commit/dab0cda47e62d8b519fcac798f9221066615d7ec)
 - [c403c52ac09373dded105ec0c5d08772730df7b6](https://github.com/magento/magento2/commit/c403c52ac09373dded105ec0c5d08772730df7b6)
 - [d7cdd7bf6f397119c83814d412ed02f8702cb7fb](https://github.com/magento/magento2/commit/d7cdd7bf6f397119c83814d412ed02f8702cb7fb)
 - [05f1a2b01bbc085b3b9db9db1217c317c1f1de42](https://github.com/magento/magento2/commit/05f1a2b01bbc085b3b9db9db1217c317c1f1de42)
 - [f25de7bffe498d35d05064814cb3671a91208c09](https://github.com/magento/magento2/commit/f25de7bffe498d35d05064814cb3671a91208c09)
 - [51a18f657bbd17155792147dd1436faa377482e6](https://github.com/magento/magento2/commit/51a18f657bbd17155792147dd1436faa377482e6)
 - [0bae5d5420efb58ac0be0abe317d01c3931e49c4](https://github.com/magento/magento2/commit/0bae5d5420efb58ac0be0abe317d01c3931e49c4)
 - [1e7082dcd43f8556e951b5b1112a14a6f29aefd7](https://github.com/magento/magento2/commit/1e7082dcd43f8556e951b5b1112a14a6f29aefd7)
 - [f5aa61dfeec3a96be19948d54eaeb701fcfc7231](https://github.com/magento/magento2/commit/f5aa61dfeec3a96be19948d54eaeb701fcfc7231)
 - [e8d3b78b81d7041d1952d4a8150a1970950a7147](https://github.com/magento/magento2/commit/e8d3b78b81d7041d1952d4a8150a1970950a7147)
 - [6c171ce75fdadddc993463a3175918ef1d2158d7](https://github.com/magento/magento2/commit/6c171ce75fdadddc993463a3175918ef1d2158d7)
 - [8c9189f8ab4dcfb82cd1c5b54ea8528aab8d8779](https://github.com/magento/magento2/commit/8c9189f8ab4dcfb82cd1c5b54ea8528aab8d8779)
 - [68cf64d2d9fb7c11b6f268ef159ab56a7db906c5](https://github.com/magento/magento2/commit/68cf64d2d9fb7c11b6f268ef159ab56a7db906c5)
 - [0db92206a634c5453780e6b9aef9fd0922f9874d](https://github.com/magento/magento2/commit/0db92206a634c5453780e6b9aef9fd0922f9874d)
 - [c44bbf5d269883a570c60304f90d8d5946ae6971](https://github.com/magento/magento2/commit/c44bbf5d269883a570c60304f90d8d5946ae6971)
 - [84fdc7a643f2039149b081dc13572a3b80e63b0b](https://github.com/magento/magento2/commit/84fdc7a643f2039149b081dc13572a3b80e63b0b)
 - [162eb14bdc3acd8a983a3633d11f2987692c7186](https://github.com/magento/magento2/commit/162eb14bdc3acd8a983a3633d11f2987692c7186)
 - [11c3eeaf6d2cd5d096aac2952eb145eaf4a2ba49](https://github.com/magento/magento2/commit/11c3eeaf6d2cd5d096aac2952eb145eaf4a2ba49)
 - [986a4ef2b1260c5c445b83f4a678f6c6af019b1d](https://github.com/magento/magento2/commit/986a4ef2b1260c5c445b83f4a678f6c6af019b1d)
 - [80cb06e283a030ac3872ca46c4be17104dd87e83](https://github.com/magento/magento2/commit/80cb06e283a030ac3872ca46c4be17104dd87e83)
 - [db71c464b0c803775c11c4988c47c4f929e32131](https://github.com/magento/magento2/commit/db71c464b0c803775c11c4988c47c4f929e32131)
 - [e105152c438dfed86918f40ec4da0f89dc537436](https://github.com/magento/magento2/commit/e105152c438dfed86918f40ec4da0f89dc537436)
 - [bd5c98ea3e094233b9a29f61a5d9082de6dd2d88](https://github.com/magento/magento2/commit/bd5c98ea3e094233b9a29f61a5d9082de6dd2d88)

The fix will be available with the upcoming 2.4.0 release.
